### ISSUE TITLE
how do you go to Eq. (4) in the paper?

### ISSUE THREAD

**Issue by JunMa11** [LIVIAETS/boundary-loss#9]
Hi, @HKervadec , thanks for the very interesting paper.
Could you make a more detailed explanation on how to go to Eq. (4) from Eq. (2) and  Eq. (3)?

Looking forward to your reply.

**Comment by JunMa11** [comment]
To be precise, I lost in the following 
> To show that (3) holds, it suffices to notice that integrating the distance function $$2\phi_{G}(p)$$ over the normal segment connecting p and and $$q_{\partial S}(p)$$ yields ||q_{\partial S(p)} - p||^2.

**Comment by HKervadec** [comment]
Hey,
Sorry for the late reply, I am somewhat on holidays, and I won't have time to reply properly until next week. 

I think the demonstration is a little bit more detailed Boykov's original paper:
```
Boykov, Yuri, et al. "An integral solution to surface evolution PDEs via geo-cuts." European Conference on Computer Vision. Springer, Berlin, Heidelberg, 2006.
```
Equation (4) in their paper. The difference in our case, is that we optimize the network weights, so some parts of the equation are constant and can be removed from the optimization problem.

I will edit this reply with more details when I am back.

**Comment by JunMa11** [comment]
Hi, @HKervadec 
Thanks for your reply. I read Yuri's ECCV-06 paper recently, but I do not understand  the similar point in the paper, at the end of Section 3.1
> Then, (1) holds because integrating the distance function 2d0(p) along a (normal) direction
connecting some point s ∈ C0 and a point q ∈ C gives ||q − s||^2 = dC_s^2. 

My question is not urgent. I'm looking forward to your reply after your vocation.
Finally, I wish you have a good holiday.

**Comment by JunMa11** [comment]
Hi @HKervadec , 
Now I figure out how Eq. (1) go to Eq. (2) based on Yuri's ECCV-06 paper. 
But I still do not figure out how to get Eq. (4) base on Eq. (1 & 2).
The question is not urgent. I'm looking forward to your reply after your vocation.

**Comment by JunMa11** [comment]
Specifically, how to convert integral domain $\Delta S$ in Eq. (3) to the combination of `S` and `G` in Eq. (4)?

**Comment by HKervadec** [comment]
Basically, $\Delta S = (S \cup G) \setminus (S \cap G)$
![$\Delta S = (S \cup G) \setminus (S \cap G)$](http://latex2png.com/output//latex_f1c5c26c6a38a64bc6aa3cec7f030478.png)
, so equation (3) can be expressed as a sum of intergrals over S, over G, minus an integral over (S union G), with some things cancelling each other.

I am going to write a detailed proof of the paper in the next few days, because some parts and approximations are not that easy to see.

Sorry for the delay for all that.

**Comment by JunMa11** [comment]
Hi @HKervadec ,
Thanks for your reply.
I'm not sure whether I get your point correctly. 
![Markdown](http://i2.bvimg.com/595387/2a879deae5755823.png)

> so equation (3) can be expressed as a sum of intergrals over S, over G, minus an integral over (S union G), with some things cancelling each other.

Following this, I still do not fighre out how Eq. (4) is derived, could you please give me more insights?

**Comment by HKervadec** [comment]
Hey,
I wrote a detailed proof. Actually, I found another way to define $\Delta S$, making the proof easier:
![proof4](https://i.imgur.com/jRvuOSU.png)
Due to your feedback, we will add an appendix in the arxiv version of the paper with that proof.


Let me know if some parts remains unclear.

**Comment by JunMa11** [comment]
Beautiful proof! I understand it.
The final loss function is \alpha L_{GD} + (1-\alpha) L_{B}.
The magnitude is [0,1] for GDice loss, while the magnitude could be more than 1000 (3D volume) for boundary loss. It seems there is a large magnitude gap between GDice and boundary loss, even with the balance coefficient \alpha. How to solve this problem?

**Comment by HKervadec** [comment]
While it is true that the magnitude of the losses can be quite different, this is easily dealt with by the use of different weights for the loss.

Also, what will mostly matter is not the value of the loss, but its derivative wrt the softmax probabilities. 
For the boundary loss, it will range between `[-image_diagonal, image_diagonal]`. A cross-entropy will be unbounded, and can go up to infinity (which will cause big stability issues). The range of the GDL derivative is trickier to find, but there is probably a few cases where the gradient can get pretty big values.

**Comment by kkhuang1990** [comment]
well right now I am trying to detect plaque (both calcified and non-calcified plaques ) from coronary artery cross-section image. Since I have annotations marked by doctors, I treated this problem as a multi-class segmentation problem (5 classes in total including background [black], central part [blue], outer part [white], calcified plaque [orange] and non-calcified plaque [red]). The input image and corresponding annotation are shown as below.
![image](https://user-images.githubusercontent.com/14921313/70391113-f4fb4f00-1a14-11ea-91ab-48917f2d5550.png)
![image](https://user-images.githubusercontent.com/14921313/70391116-fdec2080-1a14-11ea-83a3-0c6333e26ef1.png)
 I use the same loss function as you did by combining GDL loss and Surface loss under different ratios. Within the first 105 epochs, the ratio of surface loss is increasing continuously until to 1.0 and for the following 15 epochs, only the surface loss is used for training. The class-wise F1 score of training data is shown as figure below. Obviously F1 score of calcified and non-calcified plaque began to drop after the 105the epoch. Probably that's because the surface alone cannot make sure obtaining the optimal solution even when a good prior setting is given. 
![image](https://user-images.githubusercontent.com/14921313/70391182-af8b5180-1a15-11ea-80ec-7b467a1b34e9.png)
Another reason can be that most training images don't contain calcified plaque or non-calcified plaque, then it comes to another problem: how to make surface loss suitable for image-level highly imbalanced dataset. 

Looking forward to your reply.

**Comment by HKervadec** [comment]
> Within the first 105 epochs, the ratio of surface loss is increasing continuously until to 1.0 and for the following 15 epochs, only the surface loss is used for training.

Have you tried with other scheduling strategies ? Rebalancing is not the only one working (kept only this one in the conference version for space and clarity, but will have all of them in the journal version):
* Constant weight for everyone (yes it works)
* Increasing the weight of the boundary loss but keeping the regional loss constant

Or you could get creative and develop new ones specifically for your problem ; for instance involving a different weight for each class.

You can also try other regional losses, even the default cross-entropy. 

Let me know,

Hoel
